### PR TITLE
Nemo config

### LINKED ISSUE #386: Make nemo more configurable
Atm the mapping of loop variables to dimensions (longitude, ...) and other loop related values is hard-coded in nemo (see VALID_LOOP_TYPES, NEMO_LOOP_TYPE_MAPPING, NEMO_INDEX_ORDER). I (aka Mr Config) am moving those definitions into the config file (so other loop variables can be easily used, e.g. for my ROMS work). I imagine something like:
```
mapping-lon     = var: ji, start: 1, stop: jpi
...
mapping-tracers = var: jt, start: 1, stop:
```
Assuming that the config class in python allows the same key (mapping-tracer) to be used more than once. Or maybe the other way round:
```
mapping-ji = type: lon, start: 1, stop: jpi
```
The syntax above allows us to reuse existing functions, and the order of the keys would define the loop order (NEMO_INDEX_ORDERING).

Suggestions welcome :)

### PR DESCRIPTION
**Pull request by hiker**
Fixes #386: the variable name and related information are stored in the config file now, making the nemo API easy to adjust for other codes that use different variable names. E.g. in ROMS I now get lat and lon loops identified in a schedule. Note that I do not check if the values 'lat', 'lon' etc are used, as far as I can tell it should be able to change e.g. 'lat' to 'latitude' in the config file (except for the current test cases, that test for 'lat' etc.

### CODE REVIEW

File: src/psyclone/configuration.py [hunk 1]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+            if loop_type not in self._valid_loop_types:
+                valid = str(list(self._valid_loop_types.keys()))
+                raise ConfigurationError("Invalid loop type \"{0}\" found in "
+                                         "index-order in \"{1}\".\n"
+                                         "Must be one of {2}."
+                                         .format(loop_type, config.filename,
+                                                 valid))
+
+    def get_loop_type_mapping(self):
+        '''Returns the mapping of variable names to loop type.
```
Comment (arporter) [inline src/psyclone/configuration.py]: You've got a lot of "Returns" in this docstring :-) I think you can remove the first line and drop the "Returns" from the description of ":returns:"

File: src/psyclone/configuration.py [hunk 2]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+        :returns: Returns the mapping of variable names to loop type.
+        :rtype : Dictonary of strings.
```
Comment (arporter) [inline src/psyclone/configuration.py]: s/:rtype :/:rtype:/ and s/Dictonary/Dictionary/

File: src/psyclone/configuration.py [hunk 3]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+        '''
+        return self._loop_type_mapping
+
+    def get_valid_loop_types(self):
+        ''':returns: the mapping of a loop type (lon, ...) to a dictionary
```
Comment (arporter) [inline src/psyclone/configuration.py]: New-line for :returns: please plus line continuation symbols (and below for :rtype:).

Comment (arporter) [reply src/psyclone/configuration.py]: On reflection, I think this routine also needs a slightly different name as I expected it to return a list of valid loop types whereas it actually returns a map.

File: src/psyclone/configuration.py [hunk 4]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+        Example: = {"lon": {"var": "ji", "start": "1", "stop": "jpi"},
+-                    "lat": {"var": "jj", "start": "1", "stop": "jpj"} }
+
+        :rtype: dictionary with str keys, with each value being a
+        dictionary mapping 'var', 'start', and 'stop' to str.
+        '''
+        return self._valid_loop_types
+
+    def get_index_order(self):
+        ''':returns: the order in which loops should be created in
```
Comment (arporter) [inline src/psyclone/configuration.py]: New-line before :returns: plus line-continuation chars please.

File: src/psyclone/nemo.py [hunk 5]
```diff
@@ -53,23 +54,6 @@
 NEMO_SCHEDULE_COLOUR_MAP = copy.deepcopy(_BASE_CMAP)
 NEMO_SCHEDULE_COLOUR_MAP["CodeBlock"] = "red"
 
-# The valid types of loop and associated loop variable and bounds
```
Comment (arporter) [inline src/psyclone/nemo.py]: Please could you put the comments that used to be here in the Config file instead?

File: src/psyclone/nemo.py [hunk 6]
```diff
@@ -393,8 +377,9 @@ class NemoLoop(Loop, NemoFparser2ASTProcessor):
     '''
     def __init__(self, ast, parent=None):
         from fparser.two.Fortran2003 import Loop_Control
+        valid_loop_types = Config.get().api_conf("nemo").get_valid_loop_types()
         Loop.__init__(self, parent=parent,
-                      valid_loop_types=VALID_LOOP_TYPES)
+                      valid_loop_types=valid_loop_types)
```
Comment (arporter) [inline src/psyclone/nemo.py]: I think this constructor expects valid_loop_types to be a list (https://psyclone-ref.readthedocs.io/en/latest/classpsyclone_1_1psyGen_1_1Loop.html) but we're passing it a dict here (I think this was my mistake originally). We probably get away with it but please make it explicit. Maybe by altering get_valid_loop_types() to return a list (if that's all we ever use it for) or doing `valid_loop_types=list(valid_loop_types)`.

File: src/psyclone/nemo.py [hunk 7]
```diff
@@ -519,8 +506,9 @@ class NemoImplicitLoop(NemoLoop):
 
     '''
     def __init__(self, ast, parent=None):
+        valid_loop_types = Config.get().api_conf("nemo").get_valid_loop_types()
         Loop.__init__(self, parent=parent,
-                      valid_loop_types=VALID_LOOP_TYPES)
+                      valid_loop_types=valid_loop_types)
```
Comment (arporter) [inline src/psyclone/nemo.py]: Again, I think we're passing a dict as a list here.

File: src/psyclone/tests/nemo/nemo_config_test.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2018, Science and Technology Facilities Council
```
Comment (arporter) [inline src/psyclone/tests/nemo/nemo_config_test.py]: s/2018/2019/

File: src/psyclone/tests/nemo/nemo_config_test.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+from __future__ import absolute_import
+
+import pytest
+
+from psyclone.configuration import Config, ConfigurationError
+
+
+@pytest.fixture(scope="module", autouse=True)
+def setup():
```
Comment (arporter) [inline src/psyclone/tests/nemo/nemo_config_test.py]: Nice. I think you should also be "Dr. Fixture" :-)

File: src/psyclone/tests/nemo/nemo_config_test.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    '''Make sure that all tests here use nemo API, and that to clean
```
Comment (arporter) [inline src/psyclone/tests/nemo/nemo_config_test.py]: Picky but "the nemo API" and "...and that we clean-up the..."

File: src/psyclone/tests/nemo/nemo_config_test.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    the config file at the end of the tests.'''
+    Config.get().api = "nemo"
+    yield()
+    # At the end of all tests make sure that we wipe the Config object
```
Comment (arporter) [inline src/psyclone/tests/nemo/nemo_config_test.py]: Does this happen after every separate test or after all tests in this module? If the latter, is that OK or should we wipe it after every test?

File: src/psyclone/tests/nemo/nemo_config_test.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+            config = Config()
+            with pytest.raises(ConfigurationError) as err:
+                config.load(str(config_file))
+            assert "does not contain key '{0}".format(key) in str(err)
+
+    # Add an invalid index-order
+    content = _CONFIG_CONTENT + \
+        '''mapping-lon = var: i, start: 1, stop:2
+    mapping-lat = var: i, start: 1, stop:2
```
Comment (arporter) [inline src/psyclone/tests/nemo/nemo_config_test.py]: You've used the same variable name twice here ("i") which should cause an error. Please add code for this plus a test.

File: src/psyclone/tests/nemo/nemo_config_test.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    config_file = tmpdir.join("config2")
+    with config_file.open(mode="w") as new_cfg:
+        new_cfg.write(content)
+        new_cfg.close()
+
+        config = Config()
+        with pytest.raises(ConfigurationError) as err:
+            config.load(str(config_file))
+        assert "Invalid loop type \"invalid\" found " in str(err)
+        assert "Must be one of [\\'lon\\', \\'lat\\']"
```
Comment (arporter) [inline src/psyclone/tests/nemo/nemo_config_test.py]: Nice error message.

File: doc/user_guide/configuration.rst [hunk 14]
```diff
@@ -181,3 +181,45 @@ iteration-spaces        This contains definitions of additional iteration spaces
...
+This section contains configuration options that are only applicable when
+using the NEMO API.
+
+.. tabularcolumns:: |l|L|
+
+======================= =======================================================
+Entry                   Description
+======================= =======================================================
+mapping-TYPE            This declares a mapping for a certain loop level,
+                        specified as TYPE. Each value must have three key:value
```
Comment (arporter) [inline doc/user_guide/configuration.rst]: Please also document the fact that TYPE will be the "type" of any corresponding Loop objects in the PSyIR - as returned by `Loop.loop_type`

File: doc/user_guide/configuration.rst [hunk 15]
```diff
@@ -181,3 +181,45 @@ iteration-spaces        This contains definitions of additional iteration spaces
                         in the :ref:`gocean1.0-configuration` section of the
                         GOcean1.0 chapter.
 ======================= =======================================================
+
```
Comment (arporter) [inline doc/user_guide/configuration.rst]: Not your fault but earlier in this document we say "and a an optional". Please could you fix.

File: doc/user_guide/configuration.rst [hunk 16]
```diff
@@ -181,3 +181,45 @@ iteration-spaces        This contains definitions of additional iteration spaces
...
+                        ``stop``: the last loop iteration.
+
+
+index-order             Specifies the order in which loops are created when
+                        converting an implicit loop to an explicit loop.
+                        All values in this comma-separated list must have a
+                        corresponding ``mapping-TYPE`` value defined.
+======================= =======================================================
+
+Here an example of the NEMO section of a PSyclone configuration file.
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/user_guide/configuration.rst]: Done, together with explaining the example a bit more.

File: src/psyclone/configuration.py [hunk 17]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+
+
+# =============================================================================
+class NemoConfig(APISpecificConfig):
+    '''Nemo-specific Config sub-class. Holds configuration options
+    specific to the Nemo API.
+
+    :param config: The 'parent' Config object.
+    :type config: :py:class:`psyclone.configuration.Config`
+    :param section: The entry for the gocean1.0 section of \
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: Dang, and I thought I had been so careful with copy&paste ;)

File: src/psyclone/configuration.py [hunk 18]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+                    the configuration file, as produced by ConfigParser.
+    :type section:  :py:class:`configparser.SectionProxy`
+
+    '''
+
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: Done

File: src/psyclone/configuration.py [hunk 19]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+        self._index_order = []
+
+        for key in section.keys():
+            # Do not handle any keys from the DEFAULT section
+            # since they are handled by Config(), not this class.
+            if key in config.get_default_keys():
+                continue
+
+            # Handle the definition of variables
+            if key[:8] == "mapping-":
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: No, python's config class returns everything in lower case.

File: src/psyclone/configuration.py [hunk 20]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+
+        # Maps a loop type (lon, ...) to a dictionary containing the
+        # corresponding variable name and start/stop values.
+        self._valid_loop_types = {}
+
+        # The order in which loops should be created in NemoExplicitLoopTrans
+        self._index_order = []
+
+        for key in section.keys():
+            # Do not handle any keys from the DEFAULT section
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: I think we have a bit of a design problem with the config file: the DEFAULT section defines default values for variables in all (other) sections. I.e. each of the section has a distributed_memory section etc.- which we must ignore (in order to detect keys that are not valid in an API specific section).

IMHO we should rename [DEFAULT] to [SHARED], [COMMON] or so - then we wouldn't have to exclude the default keys in all of the API-specific functions :)  But that would be for a different pull request :)

File: src/psyclone/configuration.py [hunk 21]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+            # since they are handled by Config(), not this class.
+            if key in config.get_default_keys():
+                continue
+
+            # Handle the definition of variables
+            if key[:8] == "mapping-":
+                loop_type = key[8:]
+                data = self.create_dict_from_string(section[key])
+                # Make sure the required keys exist"
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: Done

File: src/psyclone/configuration.py [hunk 22]
```diff
@@ -780,3 +782,101 @@ def __init__(self, config, section):
...
+                self._loop_type_mapping[var] = loop_type
+                # And the mapping of loop type to the remaining data
+                self._valid_loop_types[loop_type] = data
+
+            elif key == "index-order":
+                self._index_order = [loop.strip() for
+                                     loop in section[key].split(",")]
+
+            else:
+                raise ConfigurationError("Invalid key \"{0}\" found in "
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: Done

File: src/psyclone/configuration.py (hunk 1 above)
Comment (hiker) [reply src/psyclone/configuration.py]: Done

File: src/psyclone/configuration.py (hunk 3 above)
Comment (hiker) [reply src/psyclone/configuration.py]: Yes, I've renamed those functions. I am not sure if I found the best names, suggestions welcome :)

File: src/psyclone/configuration.py (hunk 2 above)
Comment (hiker) [reply src/psyclone/configuration.py]: Done

File: src/psyclone/nemo.py (hunk 5 above)
Comment (hiker) [reply src/psyclone/nemo.py]: Done

File: src/psyclone/nemo.py (hunk 6 above)
Comment (hiker) [reply src/psyclone/nemo.py]: Yes, you are right. I've added a new function to the nemo config class that returns just a list.

File: src/psyclone/nemo.py (hunk 7 above)
Comment (hiker) [reply src/psyclone/nemo.py]: Done

File: src/psyclone/tests/nemo/nemo_config_test.py (hunk 8 above)
Comment (hiker) [reply src/psyclone/tests/nemo/nemo_config_test.py]: Done

File: src/psyclone/tests/nemo/nemo_config_test.py (hunk 10 above)
Comment (hiker) [reply src/psyclone/tests/nemo/nemo_config_test.py]: Done

File: src/psyclone/tests/nemo/nemo_config_test.py (hunk 11 above)
Comment (hiker) [reply src/psyclone/tests/nemo/nemo_config_test.py]: *grin* I can't be lazy, can I? The ``api="nemo" should happen only once. Resetting the config file could happen after each test that uses an invalid config file. De-facto there is only one test, so I left it all in one fixture. Even if more tests are added, as long as they use the valid config file, there is no need to reset it again, but I am happy to add a separate fixture that resets the config instance.

File: src/psyclone/tests/nemo/nemo_config_test.py (hunk 12 above)
Comment (hiker) [reply src/psyclone/tests/nemo/nemo_config_test.py]: Done (but I also changed this code to use correct/not-duplicated names.

File: doc/user_guide/configuration.rst (hunk 14 above)
Comment (hiker) [reply doc/user_guide/configuration.rst]: I've extended this section somewhat, especially added more comments to the example below. Hope that's good now.

File: doc/user_guide/configuration.rst (hunk 15 above)
Comment (hiker) [reply doc/user_guide/configuration.rst]: Done

File: doc/user_guide/configuration.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/configuration.rst]: Great, thanks.

File: src/psyclone/configuration.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/configuration.py]: :-)

File: src/psyclone/configuration.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/configuration.py]: Ah! I hadn't appreciated that before. I agree with you - please could you open another Issue?

File: src/psyclone/configuration.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/configuration.py]: Oh good.

File: src/psyclone/tests/nemo/nemo_config_test.py (hunk 11 above)
Comment (arporter) [reply src/psyclone/tests/nemo/nemo_config_test.py]: OK. We can leave it for now then.

File: src/psyclone/tests/nemo/nemo_config_test.py (hunk 12 above)
Comment (arporter) [reply src/psyclone/tests/nemo/nemo_config_test.py]: Great!

File: doc/user_guide/configuration.rst (hunk 14 above)
Comment (arporter) [reply doc/user_guide/configuration.rst]: Nice, thanks.

File: src/psyclone/nemo.py (hunk 6 above)
Comment (arporter) [reply src/psyclone/nemo.py]: Thanks.

File: src/psyclone/configuration.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: Done - #391 😁

**Comment by arporter** [comment]
pylint was all good apart from one minor error that wasn't the fault of this PR (and I've taken the liberty of fixing it). There was also one minor error when building the documentation (heading underline too short) which I've also fixed. All new code is covered and compilation tests pass.
Once Travis has completed I'll merge this to master.

**Pull request merged**